doc/design/bucketpath.rst

```diff
@@ -0,0 +1,420 @@
+
+The API's factory system compiles necessary information and provides a straightforward interface for users to create paths.
+
+
+Detailed Design
++++++++++++++++
+
+The Bucket Path API aims to align with Python's ``pathlib`` abstractions, while not mirroring its entire interface due to the extensive functionality and some aspects not being fully compatible with bucket file systems.
+The goal is to utilize common functionalities and names, to improve the ease of use and reduce the learning curve.
+
+It's important to note that wherever feasible, we adopt function and property names along with semantics from ``pathlib``.
+However, when there is a significant deviation in semantics from ``pathlib`` definitions, we choose distinct names for properties and functions. This approach ensures clarity for users regarding differences.
+
+Implementation guidelines are as follows:
+
+- Embrace and use ``pathlib`` semantics and naming conventions when applicable.
+- For significant semantic deviations, opt for unique, clear names that avoid confusion with ``pathlib`` terminology.
+
+
+.. Attention::
+
+   The subsequent subsections include code snippets intended primarily for the implementer's reference.
+   Therefore, comments and docstrings within the sudo code may need adaptation for the actual implementation.
+
+
+Pathlike
+--------
+
+.. code-block:: python
+
+    from typing import Protocol
+
+    class Pathlike(Protocol):
+
+        @property
+        def name:
+            """
+            A string representing the final path component, excluding the drive and root, if any.
+            """
+
+        @property
+        def suffix:
+            """
+            The file extension of the final component, if any.
+            """
+
+        @property
+        def root:
+            """
+            A string representing the root, if any.
+            """
+
+        @property
+        def parent:
+            """
+            The logical parent of this path.
+            """
+
+        def as_uri():
+            """
+            Represent the path as a file URI. Can be used to reconstruct the location/path.
+            """
+
+        def exists():
+            """
+            Return True if the path points to an existing file or directory.
+            """
+
+        def is_dir():
+            """
+            Return True if the path points to a directory, False if it points to another kind of file.
+            """
+
+        def is_file():
+            """
+            Return True if the path points to a regular file, False if it points to another kind of file.
+            """
+
+        def read(chunk_size: int = 8192) -> Iterable[ByteString]:
+            """
+            TBD
+            """
+
+        def write(data: ByteString | BinaryIO | Iterable[ByteString]):
+            """
+            TBD
+            """
+
+        def rm():
```

Comment:
Please note that currently there is no way to detect when a delete operation is finished, i.e. propagated to all notes of a database cluster.

Reply:
Thanks for reminding me on that, still a means of deleting files will be required.